docs: Example for XRay as a sidecar

corey-cole opened the pull request:
Update documentation to include how to use X-Ray daemon and OTEL images as sidecars.

Addresses #2931 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the Apache 2.0 License.

corey-cole committed:
docs: Example for XRay as a sidecar

iamhopaul123 commented on site/content/docs/developing/sidecars.en.md:
````diff
@@ -74,6 +74,93 @@ sidecars:
+    port: 2000/udp
+```
+
+The X-Ray sidecar needs additional IAM permissions as shown below.  Include this in addons according to the [published documentation](../developing/additional-aws-resources.en.md)
````
Maybe
```suggestion
The X-Ray sidecar needs additional IAM permissions for your task role as shown below.  Include this in addons according to the [published documentation](../developing/additional-aws-resources.en.md)
```

iamhopaul123 commented on site/content/docs/developing/sidecars.en.md:
```diff
@@ -74,6 +74,93 @@ sidecars:
+An alternative to the X-Ray daemon sidecar is the [AWS Distro for OpenTelemetry](https://aws-otel.github.io/).  This sidecar not only collects
+X-Ray trace data, but can also ship ECS metrics to third-parties such as DataDog or New Relic.
+
+To use the OpenTelemetry sidecar, first, create a valid configuration file and store it in an SSM parameter.  A standard parameter is limited to 4KB.
```
Would you mind attaching a link reference for the limit 🙏

iamhopaul123 commented on site/content/docs/developing/sidecars.en.md:
```diff
@@ -74,6 +74,93 @@ sidecars:
+receivers:
+  awsxray:
+    transport: udp
+  awsecscontainermetrics:
+  otlp:
```
Do we need to indent this? Or add content to `awsecscontainermetrics`

corey-cole replied:
I had to check, but no, all we need is the empty key.  Since it's a fair question, I will deep link to the OTEL config documentation and include a note here about the empty keys.

iamhopaul123 commented on site/content/docs/developing/sidecars.en.md:
```diff
@@ -74,6 +74,93 @@ sidecars:
+        endpoint: 0.0.0.0:4317
+
+processors:
+  batch:
```
Is it intended to leave blank?

iamhopaul123 commented on site/content/docs/developing/sidecars.en.md:
````diff
@@ -74,6 +74,93 @@ sidecars:
+      exporters: [otlp]
+```
+
+Create an SSM secret using the [copilot CLI](../commands/secret-init.en.md).  If an advanced parameter was required, it will need to be created and tagged manually.
````
```suggestion
Create an SSM secret using the [`secret init`](../commands/secret-init.en.md).  If an advanced parameter was required, it will need to be created and tagged manually.
```

corey-cole replied:
Re-reading the docs, I'm going to put this command discussion ahead of the content of the secret itself.

iamhopaul123 reviewed (commented):
Thank you so much for contributing to our sidecar docs! Only a few nits 🙏

dannyrandall commented:
Hey @corey-cole! Thanks for the contribution. Just wanted to share that we are planning on supporting X-Ray tracing for Load-Balanced Web Services in the next release, matching what [we recently launched](https://github.com/aws/copilot-cli/releases/tag/v1.17.0) for Request-Driven Web Services.

It'll look like this:
```
# manifest.yml
observability:
  tracing: awsxray
```

This should cover use cases for shipping traces to X-Ray using the [AWS Distro for OpenTelemetry](https://aws-otel.github.io/) SDK or the X-Ray SDK. However, at least in the next release, we won't have support for configuring the `aws-otel-collector` (just using the [default config](https://github.com/aws-observability/aws-otel-collector/blob/main/config.yaml)), so your example here is still super helpful for showing how to create a custom ADOT configuration.😊 

Related PR: https://github.com/aws/copilot-cli/pull/3451

corey-cole committed:
docs: Updating sidecar example according to reviewer feedback

corey-cole commented:
@dannyrandall I saw that after I wrote the docs.  Per your suggestion, I've removed the simple case as unnecessary and incorporated the content feedback.

iamhopaul123 commented on site/content/docs/developing/sidecars.en.md:
```diff
@@ -74,6 +74,81 @@ sidecars:
+
+If an advanced parameter is required, it will need to be created and tagged manually.  If the configuration fits within a standard parameter, create an SSM secret using the [`secret init`](../commands/secret-init.en.md).  The YAML document below can be used as-is with New Relic after updating the API key written as "YOUR-API-KEY-HERE".
+
+In the example YAML, the inclusion of empty keys is deliberate.  The sidecar will use the collector defaults for those keys.
```
nit:
```suggestion
In the example YAML, the inclusion of empty keys is deliberate. The sidecar will use the collector defaults for those keys.
```

iamhopaul123 commented on site/content/docs/developing/sidecars.en.md:
```diff
@@ -74,6 +74,81 @@ sidecars:
+Below is an example of running the [AWS Distro for OpenTelemetry](https://aws-otel.github.io/) sidecar with a custom configuration.  The example
+custom configuration will not only collect X-Ray trace data, but also ships ECS metrics to a third party.  The example will require an SSM secret and additional IAM permissions.
+
+To use the OpenTelemetry sidecar, first, create a valid [configuration file](https://opentelemetry.io/docs/collector/configuration/).  Next, check the size of the configuration file.  A standard parameter is [limited to 4KB](https://docs.aws.amazon.com/systems-manager/latest/APIReference/API_PutParameter.html#systemsmanager-PutParameter-request-Value).
```
```suggestion
To use the OpenTelemetry sidecar, first, create a valid [configuration file](https://opentelemetry.io/docs/collector/configuration/). Next, check the size of the configuration file.  A standard parameter is [limited to 4KB](https://docs.aws.amazon.com/systems-manager/latest/APIReference/API_PutParameter.html#systemsmanager-PutParameter-request-Value).
```

iamhopaul123 commented on site/content/docs/developing/sidecars.en.md:
````diff
@@ -74,6 +74,81 @@ sidecars:
         path: '/etc/mount1'
 ```
 
+Below is an example of running the [AWS Distro for OpenTelemetry](https://aws-otel.github.io/) sidecar with a custom configuration.  The example
````
```suggestion
Below is an example of running the [AWS Distro for OpenTelemetry](https://aws-otel.github.io/) sidecar with a custom configuration. The example
```

iamhopaul123 commented on site/content/docs/developing/sidecars.en.md:
```diff
@@ -74,6 +74,81 @@ sidecars:
+To use the OpenTelemetry sidecar, first, create a valid [configuration file](https://opentelemetry.io/docs/collector/configuration/).  Next, check the size of the configuration file.  A standard parameter is [limited to 4KB](https://docs.aws.amazon.com/systems-manager/latest/APIReference/API_PutParameter.html#systemsmanager-PutParameter-request-Value).
+If the configuration file is larger than 4K, an advanced SSM parameter must be used.
+
+If an advanced parameter is required, it will need to be created and tagged manually.  If the configuration fits within a standard parameter, create an SSM secret using the [`secret init`](../commands/secret-init.en.md).  The YAML document below can be used as-is with New Relic after updating the API key written as "YOUR-API-KEY-HERE".
```
```suggestion
If an advanced parameter is required, it will need to be created and tagged manually.  If the configuration fits within a standard parameter, create an SSM secret using the [`secret init`](../commands/secret-init.en.md). The YAML document below can be used as-is with New Relic after updating the API key written as "YOUR-API-KEY-HERE".
```

iamhopaul123 commented on site/content/docs/developing/sidecars.en.md:
````diff
@@ -74,6 +74,81 @@ sidecars:
+      exporters: [otlp]
+```
+
+Writing X-Ray traces needs additional IAM permissions as shown below.  Include this in addons according to the [published documentation](../developing/additional-aws-resources.en.md)
````
```suggestion
Writing X-Ray traces needs additional IAM permissions as shown below. Include this in addons according to the [published documentation](../developing/additional-aws-resources.en.md)
```

iamhopaul123 commented on site/content/docs/developing/sidecars.en.md:
````diff
@@ -74,6 +74,81 @@ sidecars:
 ```
 
+Below is an example of running the [AWS Distro for OpenTelemetry](https://aws-otel.github.io/) sidecar with a custom configuration.  The example
+custom configuration will not only collect X-Ray trace data, but also ships ECS metrics to a third party.  The example will require an SSM secret and additional IAM permissions.
````
```suggestion
custom configuration will not only collect X-Ray trace data, but also ship ECS metrics to a third party. The example will require an SSM secret and additional IAM permissions.
```

iamhopaul123 commented on site/content/docs/developing/sidecars.en.md:
````diff
@@ -74,6 +74,81 @@ sidecars:
+In the example YAML, the inclusion of empty keys is deliberate.  The sidecar will use the collector defaults for those keys.
+
+
+```yaml
````
A stupid question: is this example for advanced SSM parameter use case because it exceeds 4k limit? Then how about the standard one how do users specify the config for that?

corey-cole replied:
This example doesn't require the advanced SSM parameter.  I know about that solely because I tried this earlier with a complex OTEL setup that rewrote/dropped ECS metrics and that config file was like 6KB.

iamhopaul123 replied:
If it doesn't need advanced SSM parameter, does that mean users can use the local config file? If so, how do they specify the config file location?

corey-cole replied:
The local config file is required for two reasons.  One, SSM is a terrible editor for YAML :)  Two, I can't think of an easier way to check the size.  Once it's been created, size checked, and copied into SSM it can be deleted/ignored.

corey-cole committed:
Apply suggestions from code review

Co-authored-by: Penghao He <penghaoh@amazon.com>

dannyrandall reviewed (approved):
lgtm! Thanks again @corey-cole!🥰

dannyrandall committed:
Merge branch 'mainline' into mainline

gabelton commented:
@dannyrandall @iamhopaul123 do you know if it's possible to run xray for another sidecar, rather than a web service? i.e. we have a load balanced web service and it has an ip-filter flask app that runs as a sidecar and screens out requests from users not on the VPN. Currently we don't see any traces in xray console when a request is blocked by this sidecar
